src/canisters/frontend/ic-certified-assets/src/certification_types.rs

```diff
@@ -0,0 +1,137 @@
+use candid::{CandidType, Deserialize};
+use serde_cbor::ser::IoWrite;
+use serde_cbor::Serializer;
+
+use crate::{tree::NestedTree, types::AssetKey};
+
+pub type AssetHashes = NestedTree<NestedTreeKey, Vec<u8>>;
+
+#[derive(Default, Clone, Debug, CandidType, Deserialize)]
+pub struct IcCertificateExpression {
+    pub ic_certificate_expression: String,
+    /// Hash of ic_certificate_expression
+    pub expression_hash: Vec<u8>,
+}
+
+/// AssetKey that has been split into segments.
+/// E.g. `["foo", "index.html"]`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AssetPath(pub Vec<AssetKey>);
+
+/// AssetPath that is ready to be inserted into asset_hashes.
+/// E.g. `["http_expr", "foo", "index.html", "<$>", "<expr_hash>", "<response_hash>"]`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct HashTreePath(pub Vec<NestedTreeKey>);
+
+impl<T> From<T> for AssetPath
+where
+    T: AsRef<str>,
+{
+    fn from(key: T) -> Self {
+        let mut iter = key.as_ref().split("/").peekable();
+        if let Some(first_segment) = iter.peek() {
+            if *first_segment == "" {
+                iter.next();
+            }
+        }
+        Self(iter.map(|segment| segment.to_string()).collect())
+    }
+}
+
+impl AssetPath {
+    pub fn reconstruct_asset_key(&self) -> AssetKey {
+        // this reconstructs "" as "/", but this is not a problem because no http client actually requests ""
+        format!("/{}", self.0.join("/"))
+    }
+
+    pub fn asset_hash_path_v1(&self) -> HashTreePath {
+        HashTreePath(vec![
+            "http_assets".into(),
+            self.reconstruct_asset_key().into(),
+        ])
+    }
+
+    pub fn asset_hash_path_root_v2(&self) -> HashTreePath {
+        let mut hash_path: Vec<NestedTreeKey> = self
+            .0
+            .iter()
+            .map(|segment| segment.as_str().into())
+            .collect();
+        hash_path.push("<$>".into());
+        hash_path.insert(0, "http_expr".into());
+        HashTreePath(hash_path)
+    }
+}
+
+impl From<Vec<NestedTreeKey>> for HashTreePath {
+    fn from(vec: Vec<NestedTreeKey>) -> Self {
+        Self(vec)
+    }
+}
+
+impl HashTreePath {
+    pub fn as_vec(&self) -> &Vec<NestedTreeKey> {
+        &self.0
+    }
+
+    pub fn expr_path(&self) -> String {
```

Comment:
We decided to alter the spec and allow it to start with `http_expr` because it makes building the tree significantly easier. I added it as a comment to the spec since we forgot to apply the change to the spec